research/fed-bpt/src/global_es.py

```diff
@@ -0,0 +1,147 @@
+        - def run(self)
+
+    The parent classes provide the default implementations for other routines.
+
+    Args:
+        min_clients (int, optional): The minimum number of clients responses before
+            Workflow starts to wait for `wait_time_after_min_received`. Note that the workflow will move forward
+            when all available clients have responded regardless of this value. Defaults to 1000.
+        num_rounds (int, optional): The total number of training rounds. Defaults to 5.
+        persistor_id (str, optional): ID of the persistor component. Defaults to "persistor".
+        ignore_result_error (bool, optional): whether this controller can proceed if client result has errors.
+            Defaults to False.
+        allow_empty_global_weights (bool, optional): whether to allow empty global weights. Some pipelines can have
+            empty global weights at first round, such that clients start training from scratch without any global info.
+            Defaults to False.
+        task_check_period (float, optional): interval for checking status of tasks. Defaults to 0.5.
+        persist_every_n_rounds (int, optional): persist the global model every n rounds. Defaults to 1.
+            If n is 0 then no persist.
+        frac: Fraction of the number of clients used to determine the parents selection parameter. Sets popsize. Defaults to 1.
+        sigma: initial standard deviation. Defaults to 1.
+        intrinsic_dim: intrinsic dimimension of the initial solution. Defaults to 500.
+        seed: Seed for CMAEvolutionStrategy. Defaults to 42.
+        bound: bounds set for CMAEvolutionStrategy solutions if non-zero. Defaults to 0, i.e. no bounds used.
+    """
+
+    def __init__(self, *args, frac=1, sigma=1, intrinsic_dim=500, seed=42, bound=0, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.frac = frac
+        self.seed = seed
+        self.sigma = sigma
+        self.intrinsic_dim = intrinsic_dim
+        self.bound = bound
+
+    def run(self) -> None:
+        local_cma_mu = 0.0
+
+        m = max(int(self.frac * self._min_clients), 1)
+
+        self.info("Start FedBPT.")
+        cma_opts = {
+            "seed": self.seed,
+            "popsize": m,
+            "maxiter": self._num_rounds,  # args.epochs,
+            "verbose": -1,
+            "CMA_mu": m,
+        }
+        self.info(f"Start GlobalES with {cma_opts}")
+        if self.bound > 0:
+            cma_opts["bounds"] = [-1 * self.bound, 1 * self.bound]
+        global_es = cma.CMAEvolutionStrategy(self.intrinsic_dim * [0], self.sigma, inopts=cma_opts)
+
+        local_sigma_current = global_es.sigma
+
+        client_prompt_dict = {}
+        for c in range(self._min_clients):
+            client_prompt_dict[c] = [copy.deepcopy(global_es.mean)]
+        server_prompts = [copy.deepcopy(global_es.mean)]
+
+        # best_test_acc = 0
+        for self._current_round in range(self._num_rounds):
```

Comment:
In https://github.com/NVIDIA/NVFlare/pull/2475, moved these variables (`self.current_round`, `self.num_rounds`, `self.min_clients`) to BaseFedAvg and removed the underscores. 

Also they must be set as FLModel attributes, or alternatively in BaseFedAvg we could instead overwrite and wrap `send_model_and_wait()` to automatically set these for the user.

Reply:
Thanks for pointing that out. I prefer the latter option to set them in automatically in BaseFedAvg if not set by the user.

Reply:
I updated the GlobalES controller as such and filed a ticket https://github.com/NVIDIA/NVFlare/issues/2488